### PR TITLE
registration_limit check moved to ability file

### PR DESCRIPTION
**Pull request by Ana06**
Issue #955

### CODE REVIEW

**Comment by hennevogel** [comment]
How is the user ever going to know that this is denied because the registration limit is exceeded?

**Comment by Ana06** [comment]
In fact, when it is denied because the registration is not opened it is also said: `You are not authorized to access this page.` and it is also quite confusing.

I have added an error for both cases. I think that we don't need to say what of the two possibilities is causing the error as the user is redirected to root where information about registration is shown (it shows `Sorry, no places left.` or hides the register button).

I have also defined a different `can` for create, as there is a validation for that in `registration.rb`.

Also, when the user is not signed in, why is it checked that the registration is a new record?

And what happens with users with roles? I have only checked the registration limit for signed_in and not_signed_in users. Am not sure what we want to do with other users.

File: app/models/ability.rb [hunk 1]
```diff
@@ -50,7 +50,12 @@ def not_signed_in
         registration.new_record?
       end
 
-      can [:new, :create], Registration do |registration|
+      can [:new], Registration do |registration|
+        conf = registration.conference
+        conf.registration_open? && registration.new_record? && !conf.registration_limit_exceeded?
+      end
+
+      can [:create], Registration do |registration|
```
Comment (differentreality) [inline app/models/ability.rb]: We do not want to allow `create` action either, when the registration_limit is reached.

Comment (Ana06) [inline app/models/ability.rb]: But there is a validation in Registration that checks it. It won't have any effect if I add it in the ability file.

**Comment by hennevogel** [comment]
Please rebase and check @ChrisBr's changes

**Comment by Ana06** [comment]
I've rebased, but there were not merge conflicts and I am not sure of what @ChrisBr's changes I should look at.

File: app/models/ability.rb (hunk 1 above)
Comment (differentreality) [inline app/models/ability.rb]: ability.rb should still fully reflect what is happening, ie if the user has access to a specific action, or not.

And please change variable `conf` into `conference`.

File: app/controllers/conference_registrations_controller.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ class ConferenceRegistrationsController < ApplicationController
 
   def new
     @registration = Registration.new(conference_id: @conference.id)
-    authorize! :new, @registration
+    authorize! :new, @registration, message: "Sorry, you can not register for #{@conference.title}. Registration limit exceeded or the registration is not opened."
```
Comment (differentreality) [inline app/controllers/conference_registrations_controller.rb]: The registration is not **open** (instead of opened) sounds better in this case.

**Comment by hennevogel** [comment]
@Ana06 what is the status of this PR?

**Comment by Ana06** [comment]
@hennevogel you told me to check @ChrisBr's changes but I didn't find those changes. I just rebased and I don't know if I should do anything else.

**Comment by differentreality** [comment]
:+1:

**Pull request merged**